From c3c87ef92a31fc63d1a9e5cde4b0133c46769827 Mon Sep 17 00:00:00 2001
From: Thomas Sibley <trs@bestpractical.com>
Date: Thu, 2 Dec 2010 18:31:55 -0500
Subject: [PATCH] Parse <td> content with $self rather than a new instantiation

This makes sure table cell content is formatted the same way non-table
content is formatted.
---
 lib/HTML/FormatText/WithLinks/AndTables.pm |    2 +-
 t/html-formattext-withlinks-andtables.t    |   32 ++++++++++++++++++++++++++-
 2 files changed, 31 insertions(+), 3 deletions(-)

diff --git a/lib/HTML/FormatText/WithLinks/AndTables.pm b/lib/HTML/FormatText/WithLinks/AndTables.pm
index a0fa60c..be6fa80 100644
--- a/lib/HTML/FormatText/WithLinks/AndTables.pm
+++ b/lib/HTML/FormatText/WithLinks/AndTables.pm
@@ -141,7 +141,7 @@ sub _format_tables {
                     $new_tree->{_content} = [ $td ];
                     # parse the contents of the td into text
                     # this doesn't work well with nested tables...
-                    my $text = __PACKAGE__->new->_parse($new_tree);
+                    my $text = $self->_parse($new_tree);
                     # we don't want leading or tailing whitespace
                     $text =~ s/^\s+//s;
                     $text =~ s/\s+\z//s;
diff --git a/t/html-formattext-withlinks-andtables.t b/t/html-formattext-withlinks-andtables.t
index 780799f..8d33d0e 100644
--- a/t/html-formattext-withlinks-andtables.t
+++ b/t/html-formattext-withlinks-andtables.t
@@ -55,7 +55,11 @@ Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Donec hendrerit venena
 </html>|;
 my $text = HTML::FormatText::WithLinks::AndTables->convert($html, {rm=>80,cellpadding=>2});
 my $expected =
-q|   BIG
+q|   
+
+
+
+   BIG
    ===
 
    How's it going?
@@ -96,6 +100,30 @@ q|   BIG
 
 
 |;
-use Test::More tests=>1;
+use Test::More tests=>2;
 is $text, $expected,
     'HTML::FormatText::WithLinks::AndTables->convert($html,{rm=>80,cellpadding=>2})';
+
+# Test the parsing of TD content
+{
+    my $html = <<'    EOT';
+<a href="http://example.com">Link</a>
+<table><tr><td><a href="http://example.com/foo">Cell link</a></td></tr></table>
+    EOT
+    my $expected = <<'    EOT';
+ Link (http://example.com)
+
+  Cell link (http://example.com/foo)  
+
+    EOT
+    my $text = HTML::FormatText::WithLinks::AndTables->convert(
+        $html => {
+            footnote        => '',
+            after_link      => ' (%l)',
+            before_link     => '',
+            leftmargin      => 0,
+        }
+    );
+
+    is $text, $expected, "Test TD content parsing";
+}
-- 
1.7.0.4

